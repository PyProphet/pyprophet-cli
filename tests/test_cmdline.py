# encoding: utf-8
from __future__ import print_function
import pdb
import subprocess
import pytest
import os
import shutil
import collections
import tempfile

import pandas

pandas.options.display.precision=6
pandas.options.display.width=200


@pytest.fixture(scope="session")
def test_data_folder():
    here = os.path.dirname(os.path.abspath(__file__))
    return os.path.join(here, "data")


@pytest.fixture(scope="session", params=[1, 3])
def setup(test_data_folder, request):
    num_files = request.param
    tmpdir = tempfile.mkdtemp()
    print("----- tmpdir is", tmpdir)
    data_folder = os.path.join(tmpdir, "data_folder")
    os.makedirs(data_folder)
    work_folder = os.path.join(tmpdir, "work_folder")
    for i in range(num_files):
        shutil.copy(os.path.join(test_data_folder, "test_data.txt"),
                    os.path.join(data_folder, "data_%d.txt" % i))

    result_folder = os.path.join(tmpdir, "result_folder_%d" % num_files)

    Setup = collections.namedtuple("Setup", "data_folder result_folder, work_folder number_input_files")
    return Setup(data_folder, result_folder, work_folder, num_files)


def test_prepare(setup):
    cmd = ("pyprophet-cli prepare --data-folder %s --work-folder %s" % (setup.data_folder,
                                                                           setup.work_folder))
    ret_code = subprocess.call(cmd, shell=True)
    assert ret_code == 0


def test_subsample(setup):
    cmd = ("pyprophet-cli subsample --job-number 1 --job-count 1 "
           "--sample-factor 0.1 "
           "--random-seed 43 "
           "--data-filename-pattern '*.txt' "
           "--data-folder %s --work-folder %s") % (setup.data_folder, setup.work_folder)
    ret_code = subprocess.call(cmd, shell=True)
    assert ret_code == 0
    files = os.listdir(setup.work_folder)
    assert len(files) == setup.number_input_files + 1

    files.sort()
    subsamples = pandas.read_csv(os.path.join(setup.work_folder, files[-1]), sep="\t")
    assert subsamples.shape == (934, 21)

    cmd = ("pyprophet-cli subsample --job-number 1 --job-count 2 "
           "--sample-factor 0.1 "
           "--random-seed 43 "
           "--data-filename-pattern '*.txt' "
           "--data-folder %s --work-folder %s "
           "--local-folder %s") % (setup.data_folder, setup.work_folder, tempfile.mkdtemp())
    ret_code = subprocess.call(cmd, shell=True)
    assert ret_code == 0
    files = os.listdir(setup.work_folder)
    assert len(files) == 2 if setup.number_input_files == 1 else 3

    files.sort()
    subsamples = pandas.read_csv(os.path.join(setup.work_folder, files[-1]), sep="\t")
    assert subsamples.shape == (934, 21)

    cmd = ("pyprophet-cli subsample --job-number 1 --job-count 2 "
           "--sample-factor 0.1 "
           "--random-seed 43 "
           "--data-filename-pattern '*.txt' "
           "--data-folder %s --work-folder %s") % (setup.data_folder, setup.work_folder)
    ret_code = subprocess.call(cmd, shell=True)
    assert ret_code == 0
    files = os.listdir(setup.work_folder)
    assert len(files) == 2 if setup.number_input_files == 1 else 3

    files.sort()

    subsamples = pandas.read_csv(os.path.join(setup.work_folder, files[-1]), sep="\t")
    assert subsamples.shape == (934, 21)


def test_learn(setup, regtest):
    cmd = ("pyprophet-cli learn --ignore-invalid-scores --random-seed 43 --work-folder %s"
            % setup.work_folder)
    ret_code = subprocess.call(cmd, shell=True)
    assert ret_code == 0

    files = sorted(os.listdir(setup.work_folder))

    for i, file_ in enumerate(files):
        path = os.path.join(setup.work_folder, file_)
        print(i, "%7d" % os.stat(path).st_size, file_, file=regtest)

    df = pandas.read_csv(os.path.join(setup.work_folder, "weights.txt"), header=None, sep="\t")
    print(df, file=regtest)
    df = pandas.read_csv(os.path.join(setup.work_folder, "sum_stat_subsampled.txt"), sep="\t")
    print(df, file=regtest)


def test_apply_weights(setup, regtest):
    cmd = ("pyprophet-cli apply_weights --job-number 1 --job-count 1 "
           "--data-filename-pattern '*.txt' "
           "--data-folder %s --work-folder %s") % (setup.data_folder, setup.work_folder)
    ret_code = subprocess.call(cmd, shell=True)
    assert ret_code == 0


def test_score(setup, regtest):
    cmd = ("pyprophet-cli score --job-number 1 --job-count 1 "
           "--local-folder %s "
           "--data-filename-pattern '*.txt' "
           "--data-folder %s "
           "--result-folder %s "
           "--work-folder %s") % (tempfile.mkdtemp(), setup.data_folder, setup.result_folder, setup.work_folder)
    ret_code = subprocess.call(cmd, shell=True)
    assert ret_code == 0
