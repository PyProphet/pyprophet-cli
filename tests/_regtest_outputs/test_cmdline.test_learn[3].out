0    5396 full_stat_subsampled.txt
1      17 invalid_columns.txt
2     392 score_columns.txt
3  211536 subsampled_data_0.txt
4  211536 subsampled_data_1.txt
5  211536 subsampled_data_2.txt
6     795 sum_stat_subsampled.txt
7     433 weights.txt
           0
0    0.57355
1   -0.01553
2   -4.41259
3   -0.17244
4    3.59918
5   -1.37946
6    0.88409
7    0.28058
8   -0.19711
9   -0.05322
10   0.06240
11   6.14011
12   0.59689
13  -0.83396
14  24.38886
15 -14.41224
16   0.23088
   qvalue   svalue  TP  FP        TN        FN          FDR     sens   cutoff
0    0.00  0.02927   3   0  26.72222  81.27778  1.21393e-06  0.03560  7.44284
1    0.01  0.52095  44   1  25.72222  40.27778  1.11572e-02  0.52208  5.70738
2    0.02  0.59091  50   1  25.72222  34.27778  1.51581e-02  0.59328  5.63733
3    0.05  0.80541  68   4  22.72222  16.27778  5.11029e-02  0.80686  5.03281
4    0.10  1.00000  85   8  18.72222  -0.72222  8.45590e-02  1.00000  2.31468
5    0.20  1.00000  85  23   3.72222  -0.72222  2.09037e-01  1.00000  0.51744
6    0.30  1.00000  85  26   0.72222  -0.72222  2.31425e-01  1.00000 -1.25407
7    0.40      NaN NaN NaN       NaN       NaN          NaN      NaN      NaN
8    0.50      NaN NaN NaN       NaN       NaN          NaN      NaN      NaN
